### PR TITLE
Constructors with parameters in ROS2SensorComponents

### PR DESCRIPTION
**Pull request by jhanca-robotecai**
* move configuration of `ROS2GNSSSensorComponent`, `ROS2ImuSensorComponent`, `ROS2LidarSensorComponent`, and `ROS2Lidar2DSensorComponent` to separate structs
* add constructors for _ROS2SensorComponents_ that take parameters

The goal of this PR is to simplify the implementation of methods for creating sensors (it will be used in _RobotImporter_).

### CODE REVIEW

File: Gems/ROS2/Code/Source/Lidar/LidarSensorConfiguration.cpp [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+        {
+            serializeContext->Class<LidarSensorConfiguration>()
+                ->Version(1)
+                ->Field("lidarModel", &LidarSensorConfiguration::m_lidarModel)
+                ->Field("lidarImplementation", &LidarSensorConfiguration::m_lidarSystem)
+                ->Field("LidarParameters", &LidarSensorConfiguration::m_lidarParameters)
+                ->Field("IgnoreLayer", &LidarSensorConfiguration::m_ignoreLayer)
+                ->Field("IgnoredLayerIndex", &LidarSensorConfiguration::m_ignoredLayerIndex)
+                ->Field("ExcludedEntities", &LidarSensorConfiguration::m_excludedEntities)
+                ->Field("PointsAtMax", &LidarSensorConfiguration::m_addPointsAtMax);
```
Comment (adamdbrw) [inline Gems/ROS2/Code/Source/Lidar/LidarSensorConfiguration.cpp:27]: This is not related to the PR, but please raise an issue to refactor out the common configuration between Lidar2D and Lidar3D. This is a lot of repetition that is clearly visible here. Handle differences such as population of combo boxes.

**Review by adamdbrw — COMMENTED** [review]
Nice changes! How was this PR tested?

File: Gems/ROS2/Code/Source/Lidar/LidarSensorConfiguration.cpp (hunk 1 above)
Comment (jhanca-robotecai) [reply Gems/ROS2/Code/Source/Lidar/LidarSensorConfiguration.cpp:27]: Issue created: https://github.com/o3de/o3de-extras/issues/435

**Comment by jhanca-robotecai** [comment]
> Nice changes! How was this PR tested?

It was tested manually to ensure that all options are still available in _Editor_ and with some unit tests (not pushed) to test the constructors. The unit tests were based on this code: https://github.com/RobotecAI/o3de-extras/pull/10 and can be found in closed #429 for `ROS2CameraSensorComponent`. I can refactor the test code and push it to this PR if necessary.

**Comment by adamdbrw** [comment]
Test code would be great!

**Review by adamdbrw — APPROVED** [review]

File: Gems/ROS2/Code/Source/Imu/ImuSensorConfiguration.cpp [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+ *
+ */
+
+#include "ImuSensorConfiguration.h"
+#include <AzCore/Serialization/EditContext.h>
+#include <AzCore/Serialization/EditContextConstants.inl>
+
+namespace ROS2
+{
+    void ImuSensorConfiguration::Reflect(AZ::ReflectContext* context)
```
Comment (michalpelka) [inline Gems/ROS2/Code/Source/Imu/ImuSensorConfiguration.cpp:15]: I do not like that change. There are designs that use that component and this changes reflection.

File: Gems/ROS2/Code/Source/Lidar/LidarSensorConfiguration.h [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+#include <AzCore/std/string/string.h>
+
+#include "LidarRegistrarSystemComponent.h"
+#include "LidarTemplate.h"
+#include "LidarTemplateUtils.h"
+
+namespace ROS2
+{
+    //! A structure capturing configuration of a lidar sensor (to be used with ROS2LidarSensorComponent).
+    class LidarSensorConfiguration
```
Comment (michalpelka) [inline Gems/ROS2/Code/Source/Lidar/LidarSensorConfiguration.h:22]: It breaks all lidars in other gems and other projects.

**Review by michalpelka — CHANGES_REQUESTED** [review]
Change from code point of view looks nice, but it will break all prefabs using touched components.

**Review by michalpelka — COMMENTED** [review]
Also from UX point of view it looks unclean:
Pr:
image
Development:
image
Since all components were released it  should introduce translation for versions.

**Comment by adamdbrw** [comment]
@jhanca-robotecai the default view in Editor is simple to address can be fixed through an attribute (expand).
@michalpelka I believe following the pattern of configuration reflection is good. It is a shame for projects that have serialized previous versions. However, we also updated version of the Gem to 2.0 recently, since we had API breaking changes. We won't avoid doing that between versions. I believe we get interfaces into good place before the next release - and if that includes breaking changes, so be it. The next release will happen in October most likely.
@jhanca-robotecai could you make sure versions are updated in serialization of components?

**Comment by zakmat** [comment]
Change management is an inherent part of the software engineering process. The sooner we are able to battle-test our approach to component migration the better. [Version converters](https://docs.o3de.org/docs/user-guide/programming/components/reflection/serialization-context/versioning/) might be possibly used to change existing data in an unbreakable way, and developed idioms shall be kept as unit tests for further reference

**Review by adamdbrw — COMMENTED** [review]
Looks good! Does anything else need to be changed with prefabs? I see changes for ROSBot but not Proteus.

**Comment by jhanca-robotecai** [comment]
Based on the internal conversation it was decided to break the backward compatibility and fix the design issue. If decided otherwise, the sample implementation of the version converter (for one of the sensors) can be found here: https://github.com/jhanca-robotecai/o3de-extras/pull/1

The Editor's view was set to the same as before using the attributes:
image

Additionally, as the backward compatibility is broken, the parameter `visualise` was changed to `visualize`, to follow the coding guidelines (US English). 

Finally, the _prefab_ with _Slamtec_ sensor was fixed (the configuration of 3D lidar sensor was requested for 2D lidar sensor), before:
image
after:
image

**Comment by adamdbrw** [comment]
The demos need to be adjusted to development anyway, we should skip adjustments for this PR alone

**Comment by jhanca-robotecai** [comment]
> Looks good! Does anything else need to be changed with prefabs? I see changes for ROSBot but not Proteus.
There is no configuration stored in Proteus' prefab.

> The demos need to be adjusted to development anyway, we should skip adjustments for this PR alone
It is quite easy to miss the change, as de-serialization simply ignores the unknown fields without notifying the user about problems. Therefore I would strongly recommend doing it as fast as possible. E.g. `ROSConDemo` docker configuration use `development` branch of `o3de-extras` repo, which means failure right after this PR is merged.

**Comment by adamdbrw** [comment]
@michalpelka take a look please

**Review by michalpelka — APPROVED** [review]

**Pull request merged**